cosmwasm/enclaves/shared/contract-engine/src/wasm3/mod.rs

```diff
@@ -813,6 +818,24 @@ fn host_read_db(
         return Ok(ptr_to_region_in_wasm_vm as i32);
     }
 
+    let block_cache = BLOCK_CACHE.lock().unwrap();
+
+    if let Some(kv_cache) = block_cache.get(&context.contract_key) {
+        if let Some(unwrapped) = kv_cache.read(&state_key_name) {
+            debug!("Got value from cache");
+            let ptr_to_region_in_wasm_vm =
+                write_to_memory(instance, &unwrapped).map_err(|err| {
+                    debug!(
+                        "read_db() error while trying to allocate {} bytes for the value",
+                        unwrapped.len(),
+                    );
+                    err
+                })?;
+
+            return Ok(ptr_to_region_in_wasm_vm as i32);
+        }
+    }
+
```

Comment:
Wouldn't it be easier to have a single BLOCK_CACHE and read/write only from it? Seems like it just adds additional complexity to have both at the same time

Reply:
It's annoying to do, since you want to flush the cache to the state of the chain every msg (otherwise it gets messy if the cache is really large with gas and possible failures etc)

Doing it this way is a bit more complex logically, but simpler to keep track of what needs to get written to the cache